From 45a5fbd6ee03e1d10c0d418e2af646261ea49ad8 Mon Sep 17 00:00:00 2001
From: Linus Torvalds <torvalds@linux-foundation.org>
Date: Thu, 10 May 2012 12:37:10 -0700
Subject: [PATCH 026/132] vfs: move dentry name length comparison from
 dentry_cmp() into callers

All callers do want to check the dentry length, but some of them can
check the length and the hash together, so doing it in dentry_cmp() can
be counter-productive.

Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Change-Id: Ie348be750a9d6ffd9625ad3bc40d8d587716ac84
---
 fs/dcache.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/fs/dcache.c b/fs/dcache.c
index e1a6e83f0a1..99fe612f1de 100644
--- a/fs/dcache.c
+++ b/fs/dcache.c
@@ -192,9 +192,6 @@ static inline int dentry_string_cmp(const unsigned char *cs, const unsigned char
 
 static inline int dentry_cmp(const struct dentry *dentry, const unsigned char *ct, unsigned tcount)
 {
-	if (dentry->d_name.len != tcount)
-		return 1;
-
 	/*
 	 * Be careful about RCU walk racing with rename:
 	 * use ACCESS_ONCE to fetch the name pointer.
@@ -1515,6 +1512,8 @@ static struct dentry *__d_instantiate_unique(struct dentry *entry,
 			continue;
 		if (alias->d_parent != entry->d_parent)
 			continue;
+		if (alias->d_name.len != len)
+			continue;
 		if (dentry_cmp(alias, name, len))
 			continue;
 		__dget(alias);
@@ -1935,6 +1934,8 @@ seqretry:
 			}
 		}
 
+		if (dentry->d_name.len != len)
+			continue;
 		if (!dentry_cmp(dentry, str, len))
 			return dentry;
 	}
@@ -2037,6 +2038,8 @@ struct dentry *__d_lookup(const struct dentry *parent, const struct qstr *name)
 						tlen, tname, name))
 				goto next;
 		} else {
+			if (dentry->d_name.len != len)
+				goto next;
 			if (dentry_cmp(dentry, str, len))
 				goto next;
 		}
-- 
2.17.1

